llvm/lib/Support/KnownBits.cpp

```diff
@@ -232,41 +232,53 @@ KnownBits KnownBits::smin(const KnownBits &LHS, const KnownBits &RHS) {
 }
 
 KnownBits KnownBits::abdu(const KnownBits &LHS, const KnownBits &RHS) {
-  // abdu(LHS,RHS) = sub(umax(LHS,RHS), umin(LHS,RHS)).
-  KnownBits UMaxValue = umax(LHS, RHS);
-  KnownBits UMinValue = umin(LHS, RHS);
-  KnownBits MinMaxDiff = computeForAddSub(/*Add=*/false, /*NSW=*/false,
-                                          /*NUW=*/true, UMaxValue, UMinValue);
+  // If we know which argument is larger, return (sub LHS, RHS) or
+  // (sub RHS, LHS) directly.
+  if (LHS.getMinValue().uge(RHS.getMaxValue()))
+    return computeForAddSub(/*Add=*/false, /*NSW=*/false, /*NUW=*/false, LHS,
+                            RHS);
+  if (RHS.getMinValue().uge(LHS.getMaxValue()))
+    return computeForAddSub(/*Add=*/false, /*NSW=*/false, /*NUW=*/false, RHS,
+                            LHS);
```

Comment:
Incidentally, if `computeForAddSub(/*NUW=*/true)` reliably detected conflict _and_ returned a proper conflict value, then these special cases would be unnecessary. It would just fall out from the `intersectWith` case below.

Reply:
So should we provide an internal version of computeForAddSub "computeForAddSub_internal?" that doesn't clear out conflicts and have computeForAddSub wrap that?

Reply:
Dunno - that option had not occurred to me. Really I'd prefer that all of LLVM learned to embrace conflict, but that's probably a lot of work.